modules/algorithm/puzzle_15_bfs.py

```diff
@@ -0,0 +1,75 @@
+class State:
+    def __init__(self, initial_state):
+        self.initial_state = tuple(initial_state)
+
+    def __eq__(self, other):
+        return self.initial_state == other.initial_state
+    
+    def __hash__(self):
+        return hash(tuple(self.initial_state))
+    
+        
+    def generate_neighbours(self):
+        first_state = self.initial_state
+        neighbour = []
+
+        moves = {
+            'up':-3,
+            'down':+3,
+            'left':-1,
+            'right':+1
+        }
+        invalid_left = [0,3,6]
```

Comment:
This is clever, but it’s a bit hard to generalize. (E.g. imagine we wanted to extend this to a 15 puzzle or even larger).

You can keep this for now, but id also recommend storing your state as a tuple of 3-tuples. Then this invalid check is a bit easier as well!